DIPs/DIP1xxx-ajie.md

```diff
@@ -0,0 +1,246 @@
+# Foreach auto ref
+
+| Field           | Value                                                           |
+|-----------------|-----------------------------------------------------------------|
+| DIP:            | (number/id -- assigned by DIP Manager)                          |
+| Review Count:   | 0 (edited by DIP Manager)                                       |
+| Author:         | Ate Eskola (ajieskola@gmail.com)                                |
+| Implementation: | None yet                                                        |
+| Status:         | Draft                                                           |
+
+## Abstract
+
+This DIP proposes that foreach loops with `ref` keyword should only be legal when
+the elements of the range have memory addresses. It also proposes that for the current behaviour
+of `ref`, the loop can be annotated as `auto ref` instead.
+
+This is to enable the programmer to make sure that `foreach` will iterate by reference,
+while still allowing iteration over a range of non-copyable elements without explicit need
+to adapt the code for that [4].
+
+### Reference
+
+- [1] A pull request for DMD to disallow iteration by reference when the aggregate
+ does not support it:
+    * https://github.com/dlang/dmd/pull/8437
+
+- [2] Emsi-containers GitHub repository
+    * https://github.com/dlang-community/containers
+
+- [3] Meaning of rvalues and lvalues explained
+    * http://ddili.org/ders/d.en/lvalue_rvalue.html
+
+- [4] A request for this feature in bugzilla:
+    * https://issues.dlang.org/show_bug.cgi?id=4707
+
+- [5] std.algorithm.iteration.each documentation
+    * https://dlang.org/phobos/std_algorithm_iteration.html#.each
+    
+- [6] `auto ref` language specification for function return values:
+	* https://dlang.org/spec/function.html#auto-ref-functions
+
+## Contents
+* [Rationale](#rationale)
+* [Description](#description)
+* [Alternatives](#alternatives)
+* [Copyright & License](#copyright--license)
+* [Reviews](#reviews)
+
+## Rationale
+
+There are curently two ways to iterate over a range using foreach loop: By value and
+by reference.
+
+One is forced to iterate the range by reference if the element type is a `struct` with
+a disabled postblit, since iteration by value constructs the iteration variable by
+copy. As of DMD 2.087.0, you can iterate any range by reference, but there already
+was a pull request [1] to disallow such iteration for ranges whose `front` is
+an rvalue [3]. That pull request did not end up disabling such an iteration due to
+practical issues, but was approved by Andrei Alexandrescu, implying that at the concept
+of `ref` meaning iteration by any method was officially considered a weakness.
```

Comment:
What PR introduced this behavior?

Reply:
No idea, I don't even know if it has ever acted otherwise.